### DISCUSSION TITLE
Does zip.js retain executable permissions of extracted files?

### DISCUSSION THREAD

**Discussion by guest271314** [gildas-lormeau/zip.js#457] [General]
Does zip.js retain executable permissions of extracted files?

**Comment by gildas-lormeau** [comment]
zip.js has been designed to run in a browser, so it does not handle the permissions for you. 

However, you can get the raw data for the permissions via the [`Entry#externalFileAttribute`](https://gildas-lormeau.github.io/zip.js/api/interfaces/Entry.html#externalFileAttribute) property. You also need to read the [`Entry#versionMadeBy`](https://gildas-lormeau.github.io/zip.js/api/interfaces/Entry.html#versionMadeBy) property to know how to decode this raw value. Unfortunately, the specification gives no useful info to know how to achieve this, see https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.3.10.TXT (search for "external file attribute" and "version made by").

There are some info here: https://unix.stackexchange.com/questions/14705/the-zip-formats-external-file-attribute for the UNIX format. I guess the best option is to find some open-source code which implements this feature in another zip library.

**Reply by guest271314** [reply]
I have tried JSZip and your library, so far. The issues is extracting and writing the extracted files to the local file system, see https://gist.github.com/guest271314/09733eaa2de16bd79f01d41de41ce843; (TL;DR) https://groups.google.com/a/chromium.org/g/chromium-dev/c/7bb-t9RSPsA/m/UfW3lg64AgAJ.

The only way tha I have been able to preserve file permissions is to create a file having the same name first, set the permission to executable, then replace that file. Then the executable permission is preserved. That's what I do when extracting `node` executable from the Node.js nightly archive https://github.com/guest271314/download-node-nightly-executable/blob/main/index.html.

I'm trying to find a way to preserve permissions without that step, in the same way `unzip` does, using only Web API's.

**Reply by gildas-lormeau** [reply]
To my knowledge, this is unfortunately not possible in a browser by using Web API's.

**Reply by guest271314** [reply]
That is consistent with what I have found, too. I'll probably use a Native Messaging host or other means to do this.

FYI when I tested your example 

```
const helloWorldTextPromise = new Response(helloWorldStream.readable).text();
```

was throwing "Failed to fetch" error. I'm not sure if Chromium/Chrome has changed he way `Response` works re reading a `ReadableStream`. I was able to use `pipeTo(WritableStream...)` changed to the `readable`. You might want to check that.

**Reply by gildas-lormeau** [reply]
Thank you for the feedback. I cannot reproduce the "failed to fetch" issue though. https://jsfiddle.net/exnyq1ft/ runs without any errors in Chrome on my end.

**Reply by guest271314** [reply]
What happens when you do something like this

```
var {
  configure,
  BlobReader,
  ZipReader,
  ZipWriter,
} = await import("https://deno.land/x/zipjs/index.js");

var {readable, writable} = new TransformStream();

var [handle] = await showOpenFilePicker();
var blob = await handle.getFile();
var zipFileReader = new BlobReader(blob);
var zipReader = new ZipReader(zipFileReader);
var [entry] = (await zipReader.getEntries());
await entry.getData(writable);
await zipReader.close();
try {
  console.log(await new Response(readble).blob());
} catch (e) {
  console.log(e);
}
```

Is `try..catch` ever reached? Is the `Blob` logged to `console`?

**Reply by guest271314** [reply]
The typo is not the issue. This hangs on Chromium Version 120.0.6094.0 (Developer Build) (64-bit)

```
try {
  console.log(await new Response(readable).blob());
} catch (e) {
  console.log(e);
}
```

**Reply by gildas-lormeau** [reply]
I concede that the support of web streams in zip.js is far to be perfect from an API point of view. You have to run `new Response(readble).blob())` before reading the entry, keep the returned promise in a variable (e.g. `entryPromise`) and `await` this variable at the end, like in my example.

I might improve the API in a future (major) version.

**Reply by guest271314** [reply]
I was just notifying you of that case. I'm on a different OS now, Chromium Developer Build from yesterday.

I don't think `await new Response(readable).blob()` should be hanging at all, whether or not we close the `writable` side of the `TransformStream` before or after chaining a `then()` or using `await`. It shouldn't matter.

I was just notifying you of that case.

**Reply by guest271314** [reply]
I don't necessarily think this is a problem with your library. I don't think Chromium/Chrome Streams Standard implementation should be hanging just because we use `new Response(readable).blob()` after we close the `writable` side of a `TransformStream`.

**Reply by gildas-lormeau** [reply]
The real problem is `await entry.getData(writable)`, you want to call it without `await`. Thus `await new Response(readble).blob()` would not be hanging.

It's a "known" (i.e. I know it) issue related to error handling and it's my fault (I guess?). On paper, it should be possible to catch the error from the stream without using `await` and when calling `Entry#getData()` with a `WritableStream` object as parameter, thanks to `WritableStream#abort()`. In practice, I have some issues to make it work properly in all browsers because they act differently and the cause is not clear to me right now (i.e. it's probably my fault because I didn't read the manual enough).

A workaround is to call `Promise#catch()` on the promise returned by `entry.getData(writable)`, and not `await` it. See for example the piece of code below which does something similar by using `Promise.all()` in a somewhat elegant way. 
https://github.com/gildas-lormeau/zip.js/blob/2a10dda82e7961a7c73e996edf25286a23f1656b/tests/all/test-streams.js#L20-L26

**Reply by guest271314** [reply]
I havn't read your source code, yet. From a Streams developer observation it shouldn't matter where or when the `readable` side of a `TransformStream` is read. Reading `readable` should not be dependent on what the `writable` side , nor where the reading or writing occurs in the code, particularly when we have control over the both pair.

I got your gist code working to fetch Chrome-For-Testing Canary Build https://gist.github.com/guest271314/8256d618b4a94ccc368c1a61e65e4cfa. 

I'll keep trying to preserve executable permissions using only the browser. I'll first try creating the directory and setting files to size 0 named for expected executable files, set to executable, then overwrite the files, which should retain permissions.

**Reply by guest271314** [reply]
Looks like permissions are not retained using the code in the gist.

**Reply by guest271314** [reply]
For now I'm hardcoding the files that need to be set to executable

```
const executables = new Set([
  "chrome",
  "chrome-wrapper",
  "chrome_crashpad_handler",
  "chrome_sandbox",
  "libEGL.so",
  "libGLESv2.so",
  "libvk_swiftshader.so",
  "libvulkan.so.1",
  "nacl_helper",
  "nacl_helper_bootstrap",
  "nacl_irt_x86_64.nexe",
  "xdg-mime",
  "xdg-settings",
]);
// ...
async function unzip(args) {
  if (args.l) {
    await listEntries(args.l || args._[0]);
  } else {
    const archive = args._.shift();
    if (archive) {
      await unzipEntries(archive, args._);
      await Deno.remove("chrome-linux64.zip");
      for (const file of executables) {
        await Deno.chmod(`chrome-linux64/${file}`, 0o764);
      }
    }
  }
}
```
